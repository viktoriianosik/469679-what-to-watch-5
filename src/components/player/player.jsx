import React, {createRef, PureComponent} from "react";
import PropTypes from "prop-types";
import MoviePropTypes from "../movie/movie-props";
import {withRouter} from "react-router-dom";
import VideoPlayer from "../video-player/video-player";

class Player extends PureComponent {
  constructor(props) {
    super(props);

    const {match: {params: {id}}, movies} = this.props;
    this.movie = movies.find((item) => item.id === parseInt(id, 10));
    this._videoRef = createRef();
  }

  componentDidMount() {
    const src = this.movie.videoLink;
    const video = this._videoRef.current;
    video.src = src;
  }

  componentDidUpdate() {
    const {isPlayingMovie, isFullScreen} = this.props;
    const video = this._videoRef.current;

    if (isPlayingMovie) {
      video.play();
    } else {
      video.pause();
    }

    if (isFullScreen) {
      video.requestFullscreen();
    }
  }

  handleExitButtonClick(movieID) {
    this.props.history.push(`/films/${movieID}`);
  }

  render() {
    const {onPlayButtonClick, isPlayingMovie, onFullScreenButtonClick, timeElapsed, progress, onTimeUpdate} = this.props;

    return (
      <div className="player">
        <VideoPlayer isPlaying={isPlayingMovie} ref={this._videoRef} muted={false} onTimeUpdate={() => {
          onTimeUpdate(this._videoRef);
        }}/>

        <button type="button" className="player__exit" onClick={() => this.handleExitButtonClick(this.movie.id)}>Exit</button>

        <div className="player__controls">
          <div className="player__controls-row">
            <div className="player__time">
              <progress className="player__progress" value={progress} max="100"></progress>
              <div className="player__toggler" style={{left: `${progress}%`}}>Toggler</div>
            </div>
            <div className="player__time-value">{timeElapsed}</div>
          </div>

          <div className="player__controls-row">
            <button type="button" className="player__play" onClick={onPlayButtonClick}>
              <svg viewBox="0 0 19 19" width="19" height="19">
                <use xlinkHref={isPlayingMovie ? `#pause` : `#play-s`}></use>
              </svg>
              <span>{isPlayingMovie ? `Pause` : `Play`}</span>
            </button>
            <div className="player__name">{this.movie.name}</div>

            <button type="button" className="player__full-screen" onClick={onFullScreenButtonClick}>
              <svg viewBox="0 0 27 27" width="27" height="27">
                <use xlinkHref="#full-screen"></use>
              </svg>
              <span>Full screen</span>
            </button>
          </div>
        </div>
      </div>
    );
  }
}

Player.propTypes = {
  movies: PropTypes.arrayOf(MoviePropTypes).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired
    }).isRequired,
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  onPlayButtonClick: PropTypes.func.isRequired,
  onFullScreenButtonClick: PropTypes.func.isRequired,
  isPlayingMovie: PropTypes.bool.isRequired,
  isFullScreen: PropTypes.bool.isRequired,
  timeElapsed: PropTypes.string.isRequired,
  progress: PropTypes.number.isRequired,
  onTimeUpdate: PropTypes.func.isRequired,
};

export {Player};
export default withRouter(Player);
