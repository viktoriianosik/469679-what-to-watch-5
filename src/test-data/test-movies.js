export const testMovies = [
  {
    backgroundColor: `#795433`,
    backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    director: `Quentin Tarantino`,
    genre: `Crime`,
    id: 1,
    isFavorite: false,
    name: `Pulp Fiction`,
    posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 1.5,
    released: 1994,
    runTime: 153,
    scoresCount: 1635992,
    starring: [`John Travolta`, `Uma Thurman`, `Samuel L. Jackson`],
    videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  },
  {
    backgroundColor: `#BDAD8F`,
    backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/No_Country_for_Old_Men.jpg`,
    description: `Violence and mayhem ensue after a hunter stumbles upon a drug deal gone wrong and more than two million dollars in cash near the Rio Grande.`,
    director: `Ethan Coen`,
    genre: `Crime`,
    id: 14,
    isFavorite: false,
    name: `No Country for Old Men`,
    posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/No_Country_for_Old_Men.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/no-country-for-old-men.jpg`,
    previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 4.1,
    released: 2007,
    runTime: 122,
    scoresCount: 764976,
    starring: [`Tommy Lee Jones`, `Javier Bardem`, `Josh Brolin`],
    videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  }
];
