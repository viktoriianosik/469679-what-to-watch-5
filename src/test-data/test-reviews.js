export const testReviews = [
  {
    comment: `I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.`,
    date: `2020-10-09T13:38:44.769Z`,
    id: 1,
    rating: 3.3,
    user: {
      id: 19,
      name: `Christina`
    },
  },
  {
    comment: `I love this movie.`,
    date: `2020-13-09T08:38:44.769Z`,
    id: 1,
    rating: 5,
    user: {
      id: 9,
      name: `Maks`
    },
  },
  {
    comment: `Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.`,
    date: `2019-30-08T13:38:44.769Z`,
    id: 14,
    rating: 8.3,
    user: {
      id: 3,
      name: `Olga`
    },
  },
];
