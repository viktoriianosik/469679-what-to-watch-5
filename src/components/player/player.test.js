import React from "react";
import renderer from "react-test-renderer";
import {Player} from "./player";
import {testMovies} from "../../test-data/test-movies";

const noop = () => {};

it(`Render component Player correctly`, () => {
  const tree = renderer.create(
      <Player
        movies={testMovies}
        onPlayButtonClick={noop}
        onTimeUpdate={noop}
        isPlayingMovie={true}
        onFullScreenButtonClick={noop}
        isFullScreen={false}
        timeElapsed={`1h 12min`}
        progress={10}
        match={{params: {id: `1`}}}
      />
      , {
        createNodeMock: () => {
          return {};
        }
      }
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
