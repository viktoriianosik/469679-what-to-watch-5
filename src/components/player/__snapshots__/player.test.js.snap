// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render component Player correctly 1`] = `
<div
  className="player"
>
  <video
    autoPlay={true}
    className="player__video"
    muted={false}
    onTimeUpdate={[Function]}
  />
  <button
    className="player__exit"
    onClick={[Function]}
    type="button"
  >
    Exit
  </button>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max="100"
          value={10}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "10%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        1h 12min
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlinkHref="#pause"
          />
        </svg>
        <span>
          Pause
        </span>
      </button>
      <div
        className="player__name"
      >
        Pulp Fiction
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;
